File: CSE241 - Object Oriented Programming/studies/cpp/ch15-Polymorphism-and-Virtual-Functions/employee/basePlusCommissionEmployee.cpp
#include <iostream>
#include <string>
#include "basePlusCommissionEmployee.h"

using std::string;
using std::cerr;
using std::cout;

namespace GTUEmployee {
    BasePlusCommissionEmployee::BasePlusCommissionEmployee (const string & firstName, const string & lastName, const string & ssn, double grossSales, double commissionRate, double baseSalary)
    : CommissionEmployee(firstName, lastName, ssn, grossSales, commissionRate) {
        setBaseSalary(baseSalary);
    }

    void BasePlusCommissionEmployee::setBaseSalary (double baseSalary) {
        _baseSalary = (baseSalary > 0.0) ? baseSalary : 0.0;
    }   

    double BasePlusCommissionEmployee::getBaseSalary () const {return _baseSalary;} 

    double BasePlusCommissionEmployee::earnings () const {
        return getBaseSalary() + CommissionEmployee::earnings();
    } 

    void BasePlusCommissionEmployee::print() const {
        cout << "base-salaried ";
        // invoke CommissionEmployee's print function
        CommissionEmployee::print();
        cout << "\nbase salary: " << getBaseSalary();
    } 
}

File: cse241-object-oriented-programming/studies/java/inheritance/employee/BasePlusCommissionEmployee.java
package employee;

public class BasePlusCommissionEmployee extends CommissionEmployee {
    private double _baseSalary;

    public BasePlusCommissionEmployee (String fname, String lname, String ssn, double sales, double rate, double baseSalary) {
        super(fname, lname, ssn, sales, rate);
        setBaseSalary(baseSalary);    
    }

    public void setBaseSalary (double baseSalary) {_baseSalary = (baseSalary < 0.0) ? 0.0 : baseSalary;}    

    public double baseSalary () {return _baseSalary;}

    public double earnings () {return super.earnings() + baseSalary();}

    public String toString () {
        return String.format("%s %s; %s: $%,.2f", 
            "base-salalied", super.toString(), "base salary", baseSalary());
    } 
}
